lib/veres-one-did-doc.js

```diff
@@ -0,0 +1,452 @@
+   */
+  static async generate(options) {
+    const keyType = options.keyType || constants.DEFAULT_KEY_TYPE;
+    if(!constants.SUPPORTED_KEY_TYPES.includes(keyType)) {
+      throw new Error(`Unknown key type: "${keyType}"`);
+    }
+
+    const did = new VeresOneDidDoc(options);
+    await did.init({keyType, ...options});
+    return did;
+  }
+
+  /**
+   * Returns the DID uri.
+   */
+  get id() {
+    return this.doc.id;
+  }
+
+  set id(value) {
+    this.doc.id = value;
+  }
+
+  /**
+   * Initializes an empty (newly created) DID document, by generating an id,
+   * as well as authentication and authorization suites.
+   * Only called when generating a new DID Doc (creates new keys, etc).
+   *
+   * @param [env]
+   * @param [passphrase]
+   * @param [keyType] {string}
+   * @param [didType] {string}
+   *
+   * @returns {Promise}
+   */
+  async init({env, passphrase, didType, keyType = constants.DEFAULT_KEY_TYPE}) {
+    const keyOptions = {injector: this.injector, type: keyType, passphrase};
+
+    // Generate a capabilityInvocation key, to base the DID URI on
+    const invokeKey = await LDKeyPair.generate(keyOptions);
+    const did = this.generateId({keyPair: invokeKey, didType, env});
+    this.doc.id = did;
+
+    // Generate an authentication key pair and suite
+    const authKey = await LDKeyPair.generate(keyOptions);
+    authKey.id = `${did}#authn-key-1`;
+    this.doc[constants.SUITES.authentication] = [
+      this.generateSuite({key: authKey, suiteId: `${did}#auth-suite-1`})
+    ];
+    this.keys[authKey.id] = authKey;
+
+    // Generate a capabilityDelegation key pair and suite
+    const delegateKey = await LDKeyPair.generate(keyOptions);
+    delegateKey.id = `${did}#ocap-delegate-key-1`;
+    this.doc[constants.SUITES.capabilityDelegation] = [
+      this.generateSuite({key: delegateKey, suiteId: `${did}#delegate-suite-1`})
+    ];
+    this.keys[delegateKey.id] = delegateKey;
+
+    // Generate a capabilityInvocation suite (from an earlier generated key)
+    invokeKey.id = `${did}#ocap-invoke-key-1`;
+    this.doc[constants.SUITES.capabilityInvocation] = [
+      this.generateSuite({key: invokeKey, suiteId: `${did}#invoke-suite-1`})
+    ];
+    this.keys[invokeKey.id] = invokeKey;
+  }
+
+  /**
+   * Generates a DID uri, either as a globally unique random string (uuid),
+   * or from a given key pair (in case of cryptonym type did).
+   *
+   * @param [keyPair] {LDKeyPair}
+   * @param [didType='nym'] {string} 'uuid' or 'nym'. If 'nym', a key pair
+   *   must also be passed in (to generate the did uri from).
+   * @param [env='dev'] {string}
+   *
+   * @returns {string} DID uri
+   */
+  generateId({keyPair, didType = 'nym', env = 'dev'}) {
+    if(didType === 'uuid') {
+      const prefix = (env === 'live') ? 'did:v1:' : 'did:v1:test:';
+      return (prefix + 'uuid:' + uuid()).replace(/-/g, '');
+    }
+
+    if(!keyPair) {
+      throw new Error('Cannot generate a cryptonym DID without a key');
```

Comment:
Generally speaking, we typically end these types of errors with a period.

Reply:
Which types of errors? (Or rather, do you prefer ending /all/ errors with a period?)

Heh, we need a code style guide.

Reply:
@dmitrizagidulin, prefer all errors to end in a period. There are exceptions sometimes where you include a value in the error text at the end... though, we prefer those to use quotes if possible.

Reply:
Ok. (Oof, that means I have a lot of code to change across the various libs.)

Reply:
Fixed.